### PR TITLE
User Story 1415: Grouped Items in Focus Mode

### PR DESCRIPTION
**Pull request by mindlessroman**
Addressing: 
* [Focus mode excludes grouped items · Issue #165 · microsoft/vsts-extension-retrospectives (github.com)](https://github.com/microsoft/vsts-extension-retrospectives/issues/165)
* [Feature request: in focus mode, still show text of all grouped issues #176 ](https://github.com/microsoft/vsts-extension-retrospectives/issues/176)

This PR addresses a feature request for being able to see the different subitems under a grouped feedback item while in focus mode - as opposed to having to close out of the window. The sub items are shown as "Related Feedback" and limits each subitem's title to about two lines (or about 200 characters). Ideally this limits the sub-items as its general overview and not all of the title is needed.... It is an item that was deemed to lumped into another. In adjusting the way this new related feedback was styled, the timer button was moved. 

A screen shot of a feedback item. The top-level hero card has a title that says "Hello this is hannah's feedback." Beneath this hero card there is a section for adding action items and a section for 'Related Feedback." The related feedback section has three entries. Each has nonsensical gibberish typing, and one has enough characters to illustrate the two-line truncation. The aciton timer, indicating how long a feedback item has been discussed has been moved to above the card's title.

Some incidentals: 

* Adjusting how some things appear in the mobile view
* Timer button moved
* Code formatting on some files, but changing the functionality.

### CODE REVIEW

**Review by polatengin — APPROVED** [review]
omg! this is perfect! thanks :)

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx [hunk 1]
```diff
@@ -21,12 +26,18 @@ class FeedbackCarousel extends React.Component<IFeedbackCarouselProps, IFeedback
   private renderFeedbackCarouselItems = (feedbackColumnProps: FeedbackColumnProps) => {
     const columnItems = feedbackColumnProps.columnItems.sort((item1, item2) => item2.feedbackItem.upvotes - item1.feedbackItem.upvotes);
 
+    console.log('FOR THE LOVE OF CHEESE IS THE FOCUS MODAL HIDDEN? ' + this.props.isFocusModalHidden)
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: Should we remove this line?

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx [hunk 2]
```diff
@@ -356,6 +363,11 @@ class FeedbackItem extends React.Component<IFeedbackItemProps, IFeedbackItemStat
     });
   }
 
+  private toggleShowGroupedChildrenTitles = () => {
+    this.setState((previousState) => ({ isShowingGroupedChildrenTitles: !previousState.isShowingGroupedChildrenTitles }))
+
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: nit: redundant line

**Comment by phongcao** [comment]
Awesome work!

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 1 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: We should absolutely remove this line. 😂

Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: Removed !

**Review by phongcao — APPROVED** [review]

**Comment by phongcao** [comment]
@mindlessroman I don't know why the status checks don't run on your latest push. Can you try modifying something (code comment, docs...) and push to make them run again?

**Comment by mindlessroman** [comment]
> @mindlessroman I don't know why the status checks don't run on your latest push. Can you try modifying something (code comment, docs...) and push to make them run again?

Just did a minor update and that successfully triggered. Will be looking into any failures.

**Review by focusthenflex — APPROVED** [review]

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx [hunk 3]
```diff
@@ -689,7 +728,8 @@ class FeedbackItem extends React.Component<IFeedbackItemProps, IFeedbackItemStat
                   <button
                     title="UnVote"
                     aria-live="polite"
-                    aria-label={'Click to unvote On feedback. Current vote count is ' + this.props.upvotes}
+                    aria-label={'Click to unvote on feedback with title ' + this.props.title
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: minor nit, I think you may be able to do the following, but it's a matter of preference. 
```
`Click to unvote on feedback with title ${this.props.title}. `
+ `Current vote count is ${this.props.upvotes}`
```

same on line 704

**Review by nexilus18 — APPROVED** [review]

**Review by focusthenflex — APPROVED** [review]

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx [hunk 4]
```diff
@@ -68,10 +75,38 @@ class FeedbackCarousel extends React.Component<IFeedbackCarouselProps, IFeedback
...
+            if (isGroupedCarouselItem) {
+              // If an item in the column is a parent, get the title of the children
+              // to show in 'Related Feedback' section
+              const columnItemChildrenIds = columnItem.feedbackItem.childFeedbackItemIds;
+              const childrenTitlesShort: String[] = [];
+              columnItemChildrenIds.forEach(childId => {
+                // For every child item in the group, limit to 2 lines (~200 characters)
+                const childFeedbackItem = columnProps.columnItems.find(childItem => childItem.feedbackItem.id == childId)
+                const origTitle = childFeedbackItem.feedbackItem.title
+                const shortTitle = origTitle.length > 200 ? origTitle.substring(0, 200) + '...' : origTitle;
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: nit: I think `200` should be a const at the beginning of the file, making it easier to change this value in the future if the limit for child length changes.

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx [hunk 5]
```diff
@@ -68,10 +75,38 @@ class FeedbackCarousel extends React.Component<IFeedbackCarouselProps, IFeedback
         {this.props.feedbackColumnPropsList.map((columnProps) => {
           const mainCardCount = columnProps.columnItems.filter((columnItem) => !columnItem.feedbackItem.parentFeedbackItemId).length;
 
+          columnProps.columnItems.forEach(columnItem => {
+            const feedbackItemProps = FeedbackColumn.createFeedbackItemProps(columnProps, columnItem, true);
+
+            // Establish whether an item in the column has children feedback grouped beneath it,
+            // and therefore will be the parent
+            const isGroupedCarouselItem = columnItem.feedbackItem.childFeedbackItemIds ? (columnItem.feedbackItem.childFeedbackItemIds.length > 0 ? true : false) : false;
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: 🤔 I wonder if the conditional can be simplified to 

`
columnItem.feedbackItem.childFeedbackItemIds ? (columnItem.feedbackItem.childFeedbackItemIds.length > 0) : false
`

Comment (msa2984) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: same on line 86

**Review by msa2984 — APPROVED** [review]
@mindlessroman this is such a slick implementation! 💯 🎊 I left a few nits but nothing that is blocking

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx (hunk 3 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: oooooh yes I always forget about the formatting backticks! Thank you for the reminder.

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 5 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: Thank you for catching that. Ternary can be a tricky one!

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 4 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: Great idea!

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx (hunk 2 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: Good catch! Removed.

**Comment by mindlessroman** [comment]
```
Extension ID must be less than 63 characters and can only contain 'A' through 'Z', 'a' through 'z', '0' through '9', and '-'. It must start with an alphanumeric character.
Error: Process completed with exit code 255.
```

Future branches will have shorter names so this doesn't happen in the future.

**Pull request merged**